[MIG] stock_by_warehouse_sale: Migration to 17.0 T#79454

xmglord opened the pull request.

hugho-ad committed:
[ADD] stock_by_warehouse_sale, stock_by_warehouse_purchase t27399 (#1291)

moylop260 committed:
[IMP] stock_by_warehouse[_sale,_purchase]: Fix big performance overhead (#1358)

We need avoid using compute fields as possible because of the following issue:
 - https://github.com/odoo/odoo/issues/30578
warehouses_stock is a computed field used just one time but re-computed too many times:
 - Opening sale order (one time by each line)
 - Saving sale order
 - Choosing a product_id
And it computation is more and more slow when the database grows.
It is making a big overhead saving sale.order.

With this change we have the same behaviour using it just for onchange
product_id or forcing from a seudo-button (field boolean)
A real button is not used since that it requires save the record to work

Using the technical of fields with store=False and onchange methods

Check more details from https://github.com/Vauxoo/addons-vauxoo/pull/1358

moylop260 committed:
[FIX] stock_by_warehouse[_sale|_purchase]: Get quantity even if there is a warning (#1362)

moylop260 committed:
[FIX] stock_by_warehouse_sale: Using directly warehouse from order_id

In order to get the real value, since that this field is not set when onchange method is used

luisg123v committed:
[MIG] stock_by_warehouse_sale: Migrate to v12.0

The following is performed:
- Remove useless test that actually did nothing
- Remove data tags from views
- Remove encoding headers from Python files
- Remove superfluous readonly attribute from related fields
- Remove useless keys from the manifest
- Fix typos on the README

luisg123v committed:
[I18N] stock_by_warehouse_sale: Fix term names (#1392)

* [I18N] stock_by_warehouse_sale: Fix term names

Some term names are handled differently on v12.0, but those changes were
not taken into account when migrating this module from v11.0.

* [I18N] stock_by_warehouse_purchase: Fix term names

Some term names are handled differently on v12.0, but those changes were
not taken into account when migrating this module from v11.0.

luisg123v committed:
[FIX] stock_by_warehouse_sale: Remove unnecessary headers

This removes:
- Hencoding headers, fixing lint `utf8-coding-comment`
- Headers like `# Part of Odoo`
- And other ones

hugho-ad committed:
[MIG] stock_by_warehouse_sale: Migration to 14.0

luisg123v committed:
[FIX] stock_by_warehouse_sale: Edit SO line using form view

In previous Odoo versions, the form view was poped up when editing a
sale order line, as long as the option "Manage Product packaging" was
enabled. This is no longer the case [1], so we re-enable it manually,
in order to be able to see current inventory when filling a line.

[1] https://github.com/odoo/odoo/commit/e813487e75f1

rolandojduartem committed:
[MIG] stock_by_warehouse_sale: Migration to 15.0

ajqn9094 committed:
[IMP] stock_by_warehouse_sale: black, isort, prettier

rolandojduartem committed:
[REF] stock_by_warehouse_sale: Rename files by their models

rolandojduartem committed:
[MIG] stock_by_warehouse_sale: Migration to 16.0

rolandojduartem committed:
[IMP] stock_by_warehouse_sale: Make warehouses_stock visible for storable products

This functionality is for storable products to show the information in
the warehouses, now this field is invisible for products which are not
storable products.

xmglord commented:
@isaako34 and @rolandojduartem please review.

rolandojduartem commented:
@xmglord task in the MR description please

rolandojduartem commented on stock_by_warehouse_sale/i18n/es.po:
```diff
@@ -0,0 +1,46 @@
+#
+msgid ""
+msgstr ""
+"Project-Id-Version: Odoo Server 16.0+e\n"
```
Update the whole .po, please

xmglord replied:
done

rolandojduartem commented:
Please, test your PR functionally. This error happened when I tried to open a sale order line

![image](https://github.com/Vauxoo/addons-vauxoo/assets/90422721/0bccf8a9-f995-447e-9ae3-811e31639bc7)

rolandojduartem commented:
> Please, test your PR functionally. This error happened when I tried to open a sale order line
> 
> ![image](https://private-user-images.githubusercontent.com/90422721/320935779-0bccf8a9-f995-447e-9ae3-811e31639bc7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2ODIwMzMsIm5iZiI6MTcxMjY4MTczMywicGF0aCI6Ii85MDQyMjcyMS8zMjA5MzU3NzktMGJjY2Y4YTktZjk5NS00NDdlLTlhZTMtODExZTMxNjM5YmM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDE2NTUzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlOWQyMzY5MzNlMzA1NzY4MGMxOGJiYjE1ZGVlNTUxMDdkYmE0MzBjMWE0ZDYwOGNmYzlhMGJiYWZkY2QwNWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nfqd93uAvcFJdDPaQZve4kir5xH9yHdpxYzJxcixGHc)

Actually, this error comes from stock_by_warehouse, it is required functional review previous merging cc @luisg123v @desdelinux

rolandojduartem commented:
Use this as example https://github.com/odoo/odoo/blob/ce3b096de7a0dcbf2bc8387d0c17ac3d320a7ec4/addons/mail/static/src/js/onchange_on_keydown.js#L69, the fix in other PR and be sure to test in the both modules.

Commit:
[MIG] stock_by_warehouse_sale: Migration to 17.0

- Simplify the way to define modifiers as states, required, readonly,
  invisible and column_invisible as part of [1].
- Adapt the stock by warehouse field to be computed.

[1] odoo/odoo#104741

xmglord commented:
@luisg123v please review but don't merge until @rolandojduartem give his functional review.

rolandojduartem commented:
No, the changes on stock_by_warehouse must be other PR, that is a fix, not a migration, you can rebase over that PR for functional review, but I do not think is a good idea mix changes in the same PR

rolandojduartem commented on stock_by_warehouse/static/src/components/warehouse/warehouse_field.js:
```diff
@@ -70,6 +73,7 @@ export class StockByWarehouseField extends TextField {
                 title: this.info.title,
                 lines: this.lines,
                 onClose: this.closePopover,
+                show: this.show,
```
Remove this line

xmglord replied:
Done, solved in https://github.com/Vauxoo/addons-vauxoo/pull/1651

luisg123v commented on stock_by_warehouse/models/product_product.py:
```diff
@@ -19,20 +19,22 @@ class ProductProduct(models.Model):
     warehouses_stock = fields.Text(
         store=False,
         readonly=True,
+        compute="_compute_warehouses_stock",
```
Commit message:

- Missing reference
- Add comma after "finally"

xmglord replied:
Done, solved in https://github.com/Vauxoo/addons-vauxoo/pull/1651

luisg123v commented on stock_by_warehouse/models/product_product.py:
```diff
@@ -19,20 +19,22 @@ class ProductProduct(models.Model):
     )
 
-    warehouses_stock_location = fields.Text(store=False, readonly=True)
+    warehouses_stock_location = fields.Text(store=False, readonly=True, compute="_compute_warehouses_stock")
```
Since they're now computed fields, there's no need for `store=False`

xmglord replied:
Done, solved in https://github.com/Vauxoo/addons-vauxoo/pull/1651

luisg123v commented on stock_by_warehouse/static/src/components/warehouse/warehouse_field.js:
```diff
@@ -1,13 +1,13 @@
 /** @odoo-module **/
```
This is an ES module, it therefore should be named using the extension `.esm.js`.

xmglord replied:
Done, solved in https://github.com/Vauxoo/addons-vauxoo/pull/1651

luisg123v commented on stock_by_warehouse/tests/test_stock_available_unreserved.py:
```diff
@@ -147,12 +147,12 @@ def test01_stock_levels(self):
             self.assertEqual(safe_eval(product.warehouses_stock)["warehouse"], 2)
 
-        self.assertFalse(self.template_ab.warehouses_stock)
+        self.assertEqual(self.template_ab.warehouses_stock, "{}")
```
It's now possible to declare fields of type JSON, that's maybe more appropriate.

xmglord replied:
Done, solved in https://github.com/Vauxoo/addons-vauxoo/pull/1651

luisg123v commented on stock_by_warehouse_sale/models/sale_order_line.py:
```diff
@@ -4,7 +4,11 @@
+    warehouses_stock = fields.Text(
+        store=False,
+        readonly=True,
+        compute="_compute_warehouse_stock",
```
Same here, consider JSON type and store/readonly are not required.

xmglord replied:
Done

luisg123v commented on stock_by_warehouse_sale/i18n/es.po:
```diff
@@ -0,0 +1,46 @@
+#. module: stock_by_warehouse_sale
+#: model:ir.model,name:stock_by_warehouse_sale.model_sale_order_line
+msgid "Sales Order Line"
+msgstr "Línea de orden de venta"
```
Pedido

xmglord replied:
done

rolandojduartem commented:
This MR is not rebased over !1651, so it fails

rolandojduartem commented on stock_by_warehouse_sale/models/sale_order_line.py:
```diff
@@ -0,0 +1,20 @@
+
+    def _compute_get_warehouses_stock(self):
+        for line in self:
+            line.warehouses_stock = line.product_id.with_context(
```
Is it easier

```suggestion
            line.warehouses_stock = line.product_id.with_context(warehouse_id=line.warehouse_id).warehouses_stock
```

xmglord replied:
It doesn't work since the field in the product is not yet calculated and is not stored so is need to call the compute directly.

rolandojduartem replied:
True, sorry

xmglord commented:
> This MR is not rebased over !1651, so it fails

I already rebase but for some reason when installing this modules the widget fails.

desdelinux commented:
Hello, what are we missing here? @luisg123v @xmglord

rolandojduartem commented:
is this ready for functional review?

xmglord commented:
@rolandojduartem yes, this is ready for functional review.

rolandojduartem commented:
I am going to test it then

rolandojduartem commented:
I think it is not working as I expected, see the sale order, I did not activate the button and the widget was loaded at the beggining, could you check it?

![Screenshot from 2024-05-16 12-16-05](https://github.com/Vauxoo/addons-vauxoo/assets/90422721/f51007d8-a052-4785-9970-b7ed9099b188)

![Screenshot from 2024-05-16 12-14-39](https://github.com/Vauxoo/addons-vauxoo/assets/90422721/4c881840-d230-4784-942d-c2c3a4aa1645)

rolandojduartem commented on stock_by_warehouse_sale/models/sale_order_line.py:
```diff
@@ -4,20 +4,17 @@
+    def _compute_warehouse_stock(self):
+        for record in self:
+            record.warehouse_id = self.order_id.warehouse_id
+            record._compute_get_warehouses_stock()
```
Yes, you forgot the condition of warehouses_stock_recompute

xmglord replied:
done, thanks

rolandojduartem commented:
Alsoo llok the button,  the widget and the button are so close

xmglord commented:
@rolandojduartem I already fix the space between the widget and the button, please review again.

desdelinux commented:
Friendly Reminder @rolandojduartem

rolandojduartem commented:
Functionally and technically LGTM @luisg123v

luisg123v commented on stock_by_warehouse_sale/views/sale_order_views.xml:
```diff
@@ -0,0 +1,28 @@
+                        <field
+                            name="warehouses_stock_recompute"
+                            widget="boolean_toggle"
+                            options='{"autosave": False}'
```
Prefer double quotes for delimiting attributes.

luisg123v reviewed (approved):
LGTM :+1:
